README.md
sheldon-client
=====

This is the basic library to fetch data from or store data to sheldon. This
gem is not publicly available. To install it, check out the source, bundle
and install the gem manually, like that:

  git clone git@github.com:moviepilot/sheldon-client.git
  cd sheldon-client
  bundle
  rake build

Sheldon Communication
-------


As this is a simple REST wrapper around the Sheldon REST API. Please refer
to the examples and spec folder documentation and howtos.


Configuration
--------


There just two basic configuration settings. You can set the sheldon host
and you can activate logging. If you omit the :log_file option, sheldon-client
will log to stdout.

    irb > SheldonClient.host = 'http://my.sheldon/host'
    irb > SheldonClient.log  = true


Using sheldon-client
-----

Working with Nodes
---

Create a node
----
    SheldonClient.create(:node, {type: :movie, payload: {title: 'My Neighbour Totoro'}}
    => #<Sheldon::Node 204272 (Movie/My Neighbour Totoro)>


Fetch a node and set/update payload elements.
----

    SheldonClient.node 17007
    => #<Sheldon::Node 17007 (Movie/Tonari no Totoro)>

    SheldonClient.node(17007)[:title]
    => "Tonari no Totoro"

    totoro = SheldonClient.node(17007)
    totoro[:title] = "My Neighbour Totoro"
    totoro.save
    => true

    SheldonClient.update( tororo, year: '1999', title: 'My Neighbour Totoro' )
    => true

 You can also update a node without requiring the node object

    SheldonClient.update( 123, year: '1999', title: 'Matrix' )
     => true

Delete  anode
----

    totoro = SheldonClient.node 17007
    => #<Sheldon::Node 17007 (Movie/Tonari no Totoro)>

    # With the node's id
    SheldonClient.delete( node: 17007 )

    #Passing the node
    SheldonClient.delete( totoro )

Fetch connections from a node
----

    SheldonClient.node(17007).connections( :actors )
    => [ <Sheldon::Connection 64323 (Actor/17007->76423), ... ]

    SheldonClient.node(17007).connections( :actors, direction: :outgoing )
    => [ <Sheldon::Connection 64323 (Actor/17007->76423), ... ]

    SheldonClient.connection(from:17007, type: :actors)
    => [ <Sheldon::Connection 64323 (Actor/17007->76423), ... ]

With direction (incoming or outgoing)

    SheldonClient.connection(from:17007, type: :actors, direction: :outgoing)
    => [ <Sheldon::Connection 64323 (Actor/17007->76423), ... ]

Fetch connections between nodes
----

    movie    = SheldonClient.node(17007)
    person   = SheldonClient.node(96781)
    SheldonClient.connection(from:movie, to:person, type: :actings)
     => #<Sheldon::Connection 206500 (acting/17007->96781)>

    movie    = SheldonClient.node(17007)
    person   = SheldonClient.node(96781)
    SheldonClient.connection(from:movie, to:person, type: :actings)
     => #<Sheldon::Connection 206500 (acting/17007->96781)>

    SheldonClient.connection(from:17007, to:96781, type: :actings)
     => #<Sheldon::Connection 206500 (acting/17007->96781)>

Reindexing a node
-----

     gozno = SheldonClient.search( username: 'gonzo gonzales' ).first

     SheldonClient.reindex(gozno)
     => true

     heldonClient.reindex(node: gozno.id).should eq(true)
     => true

     SheldonClient.reindex(node: gozno).should eq(true)
     => true

Add/Remove a mark to a node
-------
     gonzo = SheldonClient.search( username: 'gonzo gonzales' ).first
     gonzo.mark :trending_user
     => true

     gonzo.marks
     => [:trending_user]

     gonzo.marked?
     => true

Deleting mark

     gonzo = SheldonClient.search( username: 'gonzo gonzales' ).first
     gonzo.unmark :trending_user
     => true

     gonzo.marks
     => []

     gonzo.marked?
     => false

Getting all marked_nodes
------

     marks = SheldonClient.marked_nodes
     => { buzz_bucket: [ #<Sheldon::Node 204272 (Movie/My Neighbour Totoro)>,
                         #<Sheldon::Node 204233 (Movie/Big Tits Zombie)> ],
          buzz_people: [ #<Sheldon::Node 304272 (Person/Sora Aoi)>,
                         #<Sheldon::Node 304233 (Person/Al Paccino)> ] }

Fetching all containers related with a node
----------
    node = SheldonClient.node 205352
    node.containers
    => [ #<Sheldon::Node 205352 (Container/Who Are 2011’s Oscar Contenders to Date?)>,
         #<Sheldon::Node 205352 (Container/Who Are 2011’s Oscar Contenders to Date?)> ]

Specify options

    node = SheldonClient.node 205352
    node.containers(page:1, per_page:2)
    => [ #<Sheldon::Node 205352 (Container/Who Are 2011’s Oscar Contenders to Date?)>,
         #<Sheldon::Node 205352 (Container/Who Are 2011’s Oscar Contenders to Date?)> ]

Subscribing to other node.
----------

    movie = SheldonClient.search('The Matrix').first
    user  = ShedlonClient.node 201962
    user.subscribe(movie, :everything)

Working with Connections
---

Creating connections
----

    movie = SheldonClient.node 123
    user  = SheldonClient.node 321
    SheldonClient.create :connection, { from: user,
                                        to: movie,
                                        type: :likes,
                                        payload: { weight: '0.5' } }
     => #<Sheldon::Connection 546 (likes/321->123)>

    SheldonClient.create :connection, { from: 321,
                                        to: 123,
                                        type: :likes,
                                        payload: { weight: '0.5' } }
     => #<Sheldon::Connection 545 (likes/321->123)>

Updating a connection
----

    movie = SheldonClient.node 123
    user  = SheldonClient.node 321
    connection = SheldonClient.connection(from: user, to: movie, type: :likes)

    connection[:payload] = {weight: 0.6}
    connection.save
    => true

    SheldonClient.update (:connection, { from: user,
                                        to: movie,
                                        type: :likes,
                                        payload: { weight: 0.7 } })
     => true



Deleting connections
----

    movie = SheldonClient.node 123
    user  = SheldonClient.node 321
    connection = SheldonClient.connection(from: user, to: movie, type: :likes)
    => #<Sheldon::Connection 545 (likes/321->123)>

    SheldonClient.delete connection
    => true

    SheldonClient.delete(connection: {from: 321, to: 123, type:likes})
    => true

    # Using the connection id
    SheldonClient.delete(connection: 545)
    => true

Delete all the connection of a certain type

    user  = SheldonClient.node 321
    SheldonClient.delete(connection: {from: user, type:likes})
    => true


Reindexing a connection
------

     connection = SheldonClient.high_scores(192975).first

     SheldonClient.reindex(connection)
     => true

     SheldonClient.reindex(connection: connection.id)
     => true

     SheldonClient.reindex(connection: connection)
     => true

Fetching high scores
----

    user = SheldonClient.node 192975
     => #<Sheldon::Node 192975 (User/Janis Dever Whitmer)>

    SheldonClient.high_scores user
     => [#<Sheldon::Connection 476536 (affinity/192975->191202)>]

     SheldonClient.high_scores 192975
     => [#<Sheldon::Connection 476536 (affinity/192975->191202)>]


Sheldon Status/Schema/Statistics/Ids
-------------------------

Fetching Sheldon Client Status
----------------
Returns a hash with the information under /status

     SheldonClient.status

Fetching Sheldon Client Schema
----------------
Returns a hash with information about the nodes and connections

     SheldonClient.schema

Fetching Sheldon Client Statistics
----------

     SheldonClient.statistics

Fetching ids
-----------
You can fetch the ids of nodes with a given type or all the nodes and connections,
regardless of their type.

Fetch the ids of all existing connections, regardless of their type:

     SheldonClient.all( :connections )
      => [1,2,3,4,5,6,7,8, ..... ,9999]


Fetch the ids of all existing nodes, regardless of their type:

     SheldonClient.all( :nodes )
      => [1,2,3,4,5,6,7,8, ..... ,9999]

Fetch the ids of all nodes with type movie:

     SheldonClient.all( :movies )
      => [1,2,3,4,5,6,7,8, ..... ,9999]


Searching nodes
-------

General Query
-----

     SheldonClient.search('matrix')

Will make a query to sheldon that looks like this

     http://sheldon.host:2311/search?q=matrix

Searching by parameter
----

     SheldonClient.search( facebook_ids: '501730216' )

Searches over all node and connections type that share a same parameter name


Searching by parameter with a given mode
-----
You can do exact or fulltext search.

     SheldonClient.search(title:'The Matrix', :mode => :exact )
      => [#<Sheldon::Node 416 (Movie/The Matrix)>]

     SheldonClient.search(title:'matrix', :mode => :fulltext )
      => [ #<Sheldon::Node 416 (Movie/The Matrix)>,
           #<Sheldon::Node 1658 (Movie/The Matrix Revolutions)>,
           #<Sheldon::Node 1252 (Movie/The Matrix Reloaded)>

Searching by type
-------
Searches over the sheldon's nodes of the given type like.

      SheldonClient.search(title: 'Jennifer', type: :movie)
      => [#<Sheldon::Node 4396 (Movie/Jennifer 8)>, #<Sheldon::Node 10534 (Movie/Untitled Jennifer Lopez Comedy)>]


Getting the user stream
--------------

      => gonzo = SheldonClient.search('Gonzo Gonzales').first

      => SheldonClient.stream(gonzo)
     [ #<Sheldon::Node 204272 (Container/My Neighbour Totoro and his friends news)> ]

      => SheldonClient.stream(gonzo.id)
     [ #<Sheldon::Node 204272 (Container/My Neighbour Totoro and his friends news)> ]


Doing pagerank traversals
--------------
One type of our custom traversals are pagerank like calculations on
subgraphs of sheldon, starting from a user.

    => j = SheldonClient.search('Jannis Hermanns').first
    => SheldonClient::Traverse.pagerank j, :movies
    [ { :rank => 15, :node => #<Sheldon::Node 1234 (Movie/Snatch)>},
      { :rank => 5,  :node => #<Sheldon::Node 1235 (Movie/Big Lebowski)>},
      { :rank => 1,  :node => #<Sheldon::Node 1236 (Movie/Casablanca)>} ]

Valid types are `movies`, `people`, `buckets` and `containers`. The
result list is ordered by rank, and the ranks are not normalized yet.


Batch Operations
----------
At the moment sheldon client support batch operations for connections.

      payload     = { weight: 1 }
      connections = [ { from: 13 , to: 14, type: :likes, payload: payload },
                      { from: 13 , to: 16, type: :genre_taggings, payload: payload },
                      { from: 13 , to: 20, type: :actings, payload: payload } ]


      SheldonClient.batch do |batch|
        batch.create :connection, connections[0]
        batch.create :connection, connections[1]
        batch.create :connection, connections[2]
      end



Contributing to sheldon-client
------------


* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it
* Fork the project
* Start a feature/bugfix branch
* Commit and push until you are happy with your contribution
* Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
* Please try not to mess with the Rakefile, version, or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.


Copyright
-----------

Copyright (c) 2011 moviepilot. See LICENSE.txt for further details.


CHANGELOG
=== Version 1.0.17
* Changed the way subscriber favorites json gets messed up. It gets
messed up less now
* Fixed wrong subscriber favorites response expectations seitens sheldon
client

=== Version 1.0.15
* Fixed subscriber favorites traversal url

=== Version 1.0.8
* Added support for retrieving project based on similarities recommendations
from sheldon.

=== Version 1.0.5
* Connection creation throw batch operation now operates on batches of N
  elememts each time. This is like this cause seems to be the smart way of doing
  it in neo4j.

=== Version 1.0
* Method missing support has been removed.
  Methods like node.likes and node.suggestions will raise an exception,
  use node.connections(:likes) or node.connections(:suggestions).
* Exceptions are rised for response codes different to 2.X
* SheldonClient is now a module.

=== Version 0.5.24
* Rewrite subscription action to create subscription edges. Use user_node.subscribe( other_node, :all_stories )
  * Valid subscription types are :all_stories and :featured_stories
* Added node.subscription method to fetch subscriptions for a specific node
* Fetching containers for a certain node now supports :show => :all_stories or :show => :featured_stories, defaulting to :featured_stories

=== Version 0.5.21
* Allow to specify the direction in which you want the connection.( node.connection(:likes, direction: :outgoing)
* Specify direction in which you want the neighbours.( node.neighbours(:likes, :incoming)
* Add node.subscribe method.( node.subscribe(movie_node, :everything) )

=== Version 0.5.19
* Adds support for pagerank traversal urls (SheldonClient::Traverse.pagerank)
* Adds support for node containers (SheldonClient.node(2369).containers)

=== Version 0.5.6
* Adds support for user stream. (SheldonClient.stream('23235')

=== Version 0.5
* integrate elasticsearch

=== Version 0.4

* enhancements
  * Get the neighbours directly from the node with node.neighbours
  * node.connections(:likes), Fetches all the connections of type like.
  * SheldonClient.connection, single method to fetch connections, see documentation.
  * A single method for deletion

             node = SheldonClient.search( username: 'gonzo gonzales' ).first
             SheldonClient.delete(node: node )
             SheldonClient.delete(connection: node.connections(:likes).first )
             SheldonClient.delete(connection: {from: node, type:likes})

  * Fetch statistics around data stored inside sheldon with:

             SheldonClient.statistics

  * Fetch current defined sheldon schema, includes all available nodes and connections.

             SheldonClient.schema

  * Added more examples to docs/EXAMPLES.md file.

* deprecations
  * SheldonClient.fetch_neighbours
  * SheldonClient.delete_edge ( Use SheldonClient.delete )
  * SheldonClient.edge? ( Use SheldonClient.connection(from:12, to:14, type: :like )
  * SheldonClient.edge  ( Use SheldonClient.connection(123) )
  * SheldonClient.facebook_item( '123' ) ( Use SheldonClient.search( { facebook_ids: "123" })
  * SheldonClient.get_recommendations( '123' )
  * SheldonClient.get_highscores('123') (Use SheldonClient.high_scores)
  * SheldonClient.update_edge (Use SheldonClient.update(connnection: {}, payload )
  * SheldonClient.get_node_ids_of_type ( use SheldonClient.all(:users) )
  * SheldonClient.fetch_edges ( user SheldonClient.connection )





lib/sheldon-client.rb
require 'json'
require 'active_support/inflector'
require 'forwardable'
require 'elastodon'

require 'sheldon-client/crud/crud'
require 'sheldon-client/crud/batch'
require 'sheldon-client/sheldon/status'
require 'sheldon-client/sheldon/schema'
require 'sheldon-client/sheldon/statistics'
require 'sheldon-client/sheldon/traverse'


require 'sheldon-client/configuration'
require 'sheldon-client/sheldon/sheldon_object'

module SheldonClient
  extend self
  extend SheldonClient::Configuration

  # Forward few status methods to the Status class. See
  # SheldonClient::Status for more information
  extend Forwardable
  def_delegators SheldonClient::Status, :status, :node_types, :connection_types
  def_delegators SheldonClient::Schema, :schema, :node_types, :connection_types
  def_delegators SheldonClient::Statistics, :statistics


  #
  # Create a Node or Connection in Sheldon. Please see SheldonClient::Create
  # for more information.
  #
  # === Parameters
  #
  # type    - The type of object you want to create. :node and
  #           :connection is supported.
  # options - The type-specific options. Please refer to
  #           SheldonClient::Create for more information. This
  #           should include a :type and might include a :payload
  #           element.
  #
  # ===  Examples
  #
  # Create a new node
  #
  #   SheldonClient.create :node, type: :movie, payload: { title: "Ran" }
  #
  # Create a new edge
  #
  #   SheldonClient.create :connection, type: 'like', from:    123, to:   321,
  #                         payload: { weight: 0.5 }

  def create(type, options)
    SheldonClient::Create.create_sheldon_object( type, options )
  end


  #
  # Updates the payload of a Node or Connection in Sheldon. Please see
  # SheldonClient::update for more information. Please also refer to the
  # SheldonClient::Node#update and SheldonClient::Node#[]= method.
  #
  # ==== Parameters
  #
  # object  - The object to be updated. This can be a Sheldon::Node, a
  #           Sheldon::Connection or a Hahs in the form of { <type>: <id> }
  # payload - The payload. The payload of the object will be replaces with
  #           this payload.
  #
  # ==== Examples
  #
  # Update a node
  #
  #   node = SheldonClient.node 123
  #   SheldonClient.update( node, year: '1999', title: 'Matrix' )
  #   => true
  #
  #   SheldonClient.update( { node: 123 }, title: 'Air bud' )
  #    => true
  #
  #   connection = SheldonClient.connection 123
  #   SheldonClient.update( connection, weight: 0.4 )
  #   => true
  #
  #   SheldonClient.update( { connection: {from:1, to:2, type: :like}}, weight: 0.4 )
  #    => true
  #
  def update( object, payload )
    if object.is_a?(SheldonClient::Connection)
      object = { connection: object }
    end

    SheldonClient::Update.update_sheldon_object( object, payload )
  end


  #
  # Deletes the Node or Connection from Sheldon. Please see
  # SheldonClient::Delete for more information
  #
  # ==== Parameters
  #
  # object  - The object to be updated. This can be a Sheldon::Node, a
  #           Sheldon::Connection or a Hahs in the form of { <type>: <id> }
  #
  # ==== Examples
  #
  # Delete a node from sheldon
  #
  #   SheldonClient.delete(node: 2011)
  #   => true
  #
  # Delete a connection from sheldon
  #
  #  SheldonClient.delete(connection: 201) // Non existant connection
  #   => false
  #
  # Delete a node's connections of a given type
  # SheldonClient.delete(connection: {from: 201, type: likes)
  #
  def delete( object )
    SheldonClient::Delete.delete_sheldon_object( object )
  end


  #
  # Fetch a single Node object from Sheldon. #node will return false if
  # the node could not be fetched.
  #
  # ==== Parameters
  #
  # node_id - The sheldon-id of the object to be fetched.
  #
  # ==== Examples
  #
  #   SheldonClient.node 17007
  #   => #<Sheldon::Node 17007 (Movie/Tonari no Totoro)>]
  #
  def node( node_id )
    SheldonClient::Read.fetch_sheldon_object( :node, node_id )
  end


  #
  # Search for Sheldon Nodes. This will return an array of SheldonClient::Node
  # objects or an empty array.
  #
  # ==== Parameters
  #
  # type    - plural of any known sheldon node type like :movies or :genres.
  # options - the search option that will be forwarded to lucene. This depends
  #           on the type, see below. options[:type] is reserved for the type
  #           of search you want to perform. Pass :exact or :fulltext for
  #           exact or fulltext matches.
  #
  # ==== Search Options
  #
  # Depending on the type of nodes you're searching for, different search options
  # should be provided. You can fetch the supported search keywords using the
  # status method like that:
  #
  #    SheldonClient.status['nodes']['movie']['properties'].keys
  #    => [ 'title', 'production_year', 'moviemaster_id', 'facebook_ids', ... ]
  #
  #
  # ==== Examples
  #
  # Search for a specific movie
  #
  #   SheldonClient.search :movies, { title: 'The Matrix' }
  #   SheldonClient.search :movies, { title: 'Fear and Loathing in Las Vegas', production_year: 1998 }
  #
  # Search for a specific genre
  #
  #    SheldonClient.search :genres, { name: 'Action' }
  #
  # And now with wildcards
  #
  #    SheldonClient.search :movies, { title: 'Fist*', type: fulltext }
  #

  def search( query, options = {} )
    SheldonClient::Search.search( query, options )
  end

  # Fetch a collection of edges given an url
  #
  # ==== Parameters
  #
  # * <tt> url </tt> The url where to find the edges
  #
  # ==== Examples
  #
  #  e = SheldonClient.fetch_edges("/high_scores/users/13/untracked")
  #

  def fetch_edge_collection( uri )
    fetch_collection(uri)
  end

  # Fetch a collection of edges/nodes given an url.
  #
  # ==== Parameters
  #
  # * <tt> url </tt> The url where to find the objects
  #
  # ==== Examples
  #
  #  e = SheldonClient.fetch_collection("/high_scores/users/13/untracked") # fetches edges
  #  e = SheldonClient.fetch_collection("/recommendations/users/13/containers") # fetches nodes
  #

  def fetch_collection( uri )
    SheldonClient::Read.fetch_collection( uri )
  end

  #
  # Fetches all the node ids of a given node type
  # Also all node's ids or connections's ids regardless of their type.
  #
  # === Parameters
  #
  #   * <tt>type</tt> - The node type
  #
  # === Examples
  #
  #   SheldonClient.all( :movies )
  #   => [1,2,3,4,5,6,7,8, ..... ,9999]
  #
  #   SheldonClient.all( :nodes )
  #   => [1,2,3,4,5,6,7,8, ..... ,9999]
  #
  #   SheldonClient.all( :connections )
  #   => [1,2,3,4,5,6,7,8, ..... ,9999]
  def all( type )
    SheldonClient::Read.fetch_node_type_ids(type)
  end


  #
  # Reindex an edge in Sheldon
  #
  # === Parameters
  #
  #  * <tt> edge_id </tt>
  #
  # === Examples
  #
  # SheldonClient.reindex_edge( 5464 )
  #

  def reindex( object )
    SheldonClient::Update.reindex(object)
  end

  #
  # Fetches a connection
  #
  # === Parameters
  #
  # * object -  Can be the connection id or a hash specifyin from, to, and the type.
  #
  #
  # === Examples
  #
  # from = SheldonClient.search({title: 'The Matrix} )
  # to = SheldonClient.search({name: 'Action'})
  # connection = SheldonClient.connection( from:from, to:to, type: :genres )
  # => #<Sheldon::Connection 5 (GenreTagging/1->2)>
  #
  # Passing the connection id
  #
  # SheldonClient.connection 5
  # => #<Sheldon::Connection 5 (GenreTagging/1->2)>
  #
  # SheldonClient.connection (from: 1, type: :genres)
  #
  # => [#<Sheldon::Connection 5 (GenreTagging/1->2)>, #<Sheldon::Connection 6 (GenreTagging/1->3)>]
  #
  def connection(object)
    SheldonClient::Read.fetch_sheldon_connection(object)
  end

  class << self
    alias :connections :connection
  end

  #
  # temporarily set a different host to connect to. This
  # takes a block where the given sheldon node should be
  # the one we're talking to
  #
  # == Parameters
  #
  # <tt>host</tt> - The sheldon-host (including http)
  # <tt>block</tt> - The block that should be executed
  #
  # == Examples
  #
  # SheldonClient.with_host( "http://www.sheldon.com" ) do
  #   SheldonClient.node( 1234 )
  # end
  def with_host( host, &block )
    begin
      SheldonClient.temp_host = host
      yield
    ensure
      SheldonClient.temp_host = nil
    end
  end

  # Fetches all the marks in sheldon and the marked nodes.
  #
  #  marks = SheldonClient.marked_nodes
  #  => { buzz_bucket: [ #<Sheldon::Node 204272 (Movie/My Neighbour Totoro)>,
  #                      #<Sheldon::Node 204233 (Movie/Big Tits Zombie)> ],
  #       buzz_people: [ #<Sheldon::Node 304272 (Person/Sora Aoi)>,
  #                      #<Sheldon::Node 304233 (Person/Al Paccino)> ] }

  def marked_nodes
    nodes = search(marked: true)
    marks = Hash.new do |h,k|
        h[k] = []
    end
    nodes.each do |node|
      node.marks.each do |mark|
        marks[mark] << node
      end
    end
    marks
  end

  # Returns a list of containers related to a given user, includes likes,
  # with the newest at the begining of the list.
  # You also can paginate the request using the per_page and page options.
  #
  # == Parameters
  #
  # <tt>user_id</tt> - You can pass an user object or the user id.
  # <tt>options</tt> - You can specify page and per_page.
  #
  # == Examples
  #
  #  => gonzo = SheldonClient.search('Gonzo Gonzales').first
  #
  #  => SheldonClient.stream(gonzo)
  # [ #<Sheldon::Node 204272 (Container/My Neighbour Totoro and his friends news)> ]
  #
  #  => SheldonClient.stream(gonzo.id)
  # [ #<Sheldon::Node 204272 (Container/My Neighbour Totoro and his friends news)> ]
  #
  #  => SheldonClient.stream(gonzo, page: 5, per_page: 5)
  # [ #<Sheldon::Node 204272 (Container/My Neighbour Totoro and his friends news)> ]
  #
  def stream(node, options = {})
    SheldonClient::Read.get_stream(node, options)
  end

  # Process the given block as a batch operation in sheldon
  #
  # == Parameters
  # <tt> size   </tt> - Lets you specify the size of the batch operation. By default it
  #                     is of 50 elements, so if you have more two or more request will
  #                     be issued.
  # <tt> block  </tt> - Receives a block which takes a batch as argument, in the batch
  #                     you can call create, with the type and info.
  #
  # == Example
  #
  # payload     = { weight: 1 }
  # connections = [ { from: 13 , to: 14, type: :likes, payload: payload },
  #                 { from: 13 , to: 16, type: :genre_taggings, payload: payload },
  #                 { from: 13 , to: 20, type: :actings, payload: payload } ]
  #
  # SheldonClient.batch do |batch|
  #   batch.create :connection, connections[0]
  #   batch.create :connection, connections[1]
  #   batch.create :connection, connections[2]
  # end
  #
  # SheldonClient.batch(2) do |batch|
  #   batch.create :connection, connections[0]
  #   batch.create :connection, connections[1]
  #   batch.create :connection, connections[2]
  # end
  #
  # Note
  # At the moment the creation in batch is just supported for connections.
  #
  def batch(size=50, &block)
    SheldonClient::Create.batch(size, &block)
  end

  def questionnaire(id)
    SheldonClient::Read.questionnaire id
  end

  # Special method: Retrieves the list of all nodes, identified by their neo4j ids. It's important to 
  # notices that cause this ids are only in use inside the special methods, not on the
  # regular ones.
  #
  def all_nodes
    SheldonClient::Read.all_nodes
  end

  # Special method: Retrieves the list of all connections, identified by their neo4j ids. It's
  # important to notices that cause this ids are only in use inside the special 
  # methods, not on the regular ones.
  #
  # == Parameters
  # <tt> type <tt> - The connection type we want to get the connection from.
  #
  def all_connections(type)
    SheldonClient::Read.all_connections(type)
  end

  # Special method: It gets the 1000 newest containers from sheldon
  #
  def newest_containers
    SheldonClient::Read.newest_containers
  end

  # Special method: It repair a node, attaching it to their rules, fixing external_ids, etc..
  # so it takes care that a node is consistent again.
  #
  def repair_node(id)
    SheldonClient::Update.repair_node(id)
  end

  # Special method: It repair a connection, attaching it to their rules, 
  # fixing external_ids, etc.. so it takes care that a connection is consistent again.
  #
  def repair_connection(id)
    SheldonClient::Update.repair_connection(id)
  end

  # Special method: It takes care of initializing the connection schema nodes in order
  # to be sure all the proper counters are initialized before we start to repair 
  # connections.
  #
  def initialize_connections_rules
    SheldonClient::Update.initialize_connections_rules
  end

  ## Special method: Get's a list of nodes that belongs to a certain rule.
  #
  def all_nodes_in_a_rule(rule)
    SheldonClient::Read.all_nodes_in_a_rule(rule)
  end

  def subscriber_favorites(options={})
    SheldonClient::Read.get_global_subscriber_favorites(options)
  end

  ##
  # Fetch the user activity from sheldon
  #
  def activity(id)
    SheldonClient::Read.activities id
  end

end
